### ISSUE TITLE
Build and publish against current Scala 2.12 milestone

### ISSUE THREAD

**Issue by JoshRosen** [twitter/chill#252]
It would be great if Chill could publish a version of its next release which is built against the current Scala 2.12 milestone (currently 2.12.0-M4). This is a blocker for building Apache Spark against Scala 2.12.

**Comment by JoshRosen** [comment]
I did an initial investigation of this and it's going to be slightly non-trivial; my WIP build changes are at https://github.com/twitter/chill/compare/develop...JoshRosen:scala-2.12

`chill-scala` compiles fine but several tests are failing due to issues with Java 8 lambdas. For example (`sbt ++2.12.0-M3 clean chill/test`):

``` scala
[info] - should Handle Ordering.reverse *** FAILED ***
[info]   com.esotericsoftware.kryo.KryoException: Unable to find class: com.twitter.chill.KryoSpec$$Lambda$3401/1855666130
[info] Serialization trace:
[info] cmp$1 (scala.math.Ordering$$anon$9)
[info] $outer (scala.math.Ordering$$anon$4)
[info]   at com.esotericsoftware.kryo.util.DefaultClassResolver.readName(DefaultClassResolver.java:156)
[info]   at com.esotericsoftware.kryo.util.DefaultClassResolver.readClass(DefaultClassResolver.java:133)
[info]   at com.esotericsoftware.kryo.Kryo.readClass(Kryo.java:670)
[info]   at com.esotericsoftware.kryo.serializers.ObjectField.read(ObjectField.java:118)
[info]   at com.esotericsoftware.kryo.serializers.FieldSerializer.read(FieldSerializer.java:551)
[info]   at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:708)
[info]   at com.esotericsoftware.kryo.serializers.ObjectField.read(ObjectField.java:125)
[info]   at com.esotericsoftware.kryo.serializers.FieldSerializer.read(FieldSerializer.java:551)
[info]   at com.esotericsoftware.kryo.Kryo.readClassAndObject(Kryo.java:790)
[info]   at com.twitter.chill.java.PriorityQueueSerializer.read(PriorityQueueSerializer.java:67)
[info]   at com.twitter.chill.java.PriorityQueueSerializer.read(PriorityQueueSerializer.java:32)
[info]   at com.esotericsoftware.kryo.Kryo.readClassAndObject(Kryo.java:790)
[info]   at com.twitter.chill.SerDeState.readClassAndObject(SerDeState.java:61)
[info]   at com.twitter.chill.KryoPool.fromBytes(KryoPool.java:94)
[info]   at com.twitter.chill.BaseProperties$class.deserialize(BaseProperties.scala:25)
[info]   at com.twitter.chill.KryoSpec.deserialize(KryoSpec.scala:55)
[info]   at com.twitter.chill.BaseProperties$class.rt(BaseProperties.scala:27)
[info]   at com.twitter.chill.KryoSpec.rt(KryoSpec.scala:55)
[info]   at com.twitter.chill.KryoSpec.com$twitter$chill$KryoSpec$$$anonfun$38(KryoSpec.scala:299)
[info]   at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
[info]   at org.scalatest.Transformer.org$scalatest$Transformer$$$anonfun$1(Transformer.scala:22)
```

This looks related to https://github.com/EsotericSoftware/kryo/issues/215. I haven't found an easy fix yet, but also haven't spent tons of time looking into this.

Additionally, publishing the `chill-akka`, `chill-algebird`, and `chill-bijection` libraries will require Akka, Algebird, and Bijection to publish 2.12 releases. For my own purposes in Spark, I only care about `chill-scala` and therefore left this out of scope for now. Via some SBT magic, it might be possible to disable those subprojects when building for 2.12.

**Comment by johnynek** [comment]
we can easily publish just chill-scala, that requires no magic (you can publish subprojects independently).

It looks like, we would need to register their ClosureSerializer, but even then it looks like this could present some real issues for function serialization. It seems they are only able to use java serialization for all then data, kryo cannot be used for any of the captured objects, unless I misread.

If that is true, it will be an enormous barrier for scalding, spark and other systems that use kryo to serialize closures if we can't find a solution.

**Comment by steveloughran** [comment]
One thing to consider with closure ser/deser is "does it introduce any more risks to serialization". Groovy's closure deserialization did, but that may be due to how it is more interpreted than java/scala.

**Comment by JoshRosen** [comment]
Spark currently supports Kryo as a data serializer for data transfer and storage.

Spark does not currently use Kryo as a closure serializer due to some bugs that we ran into in Kryo 2 (see https://issues.apache.org/jira/browse/SPARK-7708) but it would be nice to support that since I suspect it could have a decent performance impact for certain complex closures (e.g. serializing expression trees in Spark SQL).

Even if our current use of Kryo tends to be limited to data rather than closures, I was hoping that we'd be able to get all of Chill's automated tests to pass under Scala 2.12 so that the build configuration that we use actually gets tested as part of Chill's development process.

**Comment by steveloughran** [comment]
I see the need for this stuff within spark, just need to make sure that kryo can't be used to deserialze closures stored in files, the way `SparkContext.objectFile()` can do with java-serialized groovy closures. Support that and privilege escalation is a matter of getting your target to load a file you've created. 

Between services, the shared secrets &c make the risk less significant.

**Comment by JoshRosen** [comment]
@steveloughran, I think the discussion of possible security impacts of Kryo serialization is orthogonal to this particular issue, so I suggest that we fork off that discussion into a new thread or JIRA if you'd like to continue it.

**Comment by JoshRosen** [comment]
Note: now that 2.12.0-M4 is out we should compile against that and not M3 because apparently there are some binary incompatibilities between the two milestones.

**Comment by JoshRosen** [comment]
Now that https://github.com/EsotericSoftware/kryo/pull/415 was merged, I tried this again and got a little further: https://github.com/twitter/chill/compare/develop...JoshRosen:scala-2.12

It's now failing a test in the `ClosureCleanerSpec`:

``` scala
[info] ClosureCleaner
[info] - should clean normal objects
[info] - should clean actual closures *** FAILED ***
[info]   java.lang.RuntimeException:
[info]   at com.esotericsoftware.reflectasm.shaded.org.objectweb.asm.MethodVisitor.visitParameter(Unknown Source)
[info]   at com.esotericsoftware.reflectasm.shaded.org.objectweb.asm.ClassReader.b(Unknown Source)
[info]   at com.esotericsoftware.reflectasm.shaded.org.objectweb.asm.ClassReader.accept(Unknown Source)
[info]   at com.esotericsoftware.reflectasm.shaded.org.objectweb.asm.ClassReader.accept(Unknown Source)
[info]   at com.twitter.chill.ClosureCleaner$.getInnerClassesFn(ClosureCleaner.scala:115)
[info]   at com.twitter.chill.ClosureCleaner$.com$twitter$chill$ClosureCleaner$$$anonfun$5(ClosureCleaner.scala:125)
[info]   at scala.collection.mutable.AbstractMap.getOrElseUpdate(Map.scala:80)
[info]   at com.twitter.chill.ClosureCleaner$.innerClassesOf(ClosureCleaner.scala:125)
[info]   at com.twitter.chill.ClosureCleaner$.getAccessedFields(ClosureCleaner.scala:133)
[info]   at com.twitter.chill.ClosureCleaner$.accessedFieldsOf(ClosureCleaner.scala:147)
[info]   at com.twitter.chill.ClosureCleaner$.com$twitter$chill$ClosureCleaner$$$anonfun$12(ClosureCleaner.scala:188)
[info]   at scala.collection.immutable.List.foldLeft(List.scala:84)
[info]   at com.twitter.chill.ClosureCleaner$.allocCleanedOuter(ClosureCleaner.scala:177)
[info]   at com.twitter.chill.ClosureCleaner$.apply(ClosureCleaner.scala:162)
[info]   at com.twitter.chill.ClosureCleanerSpec.com$twitter$chill$ClosureCleanerSpec$$$anonfun$4(ClosureCleanerSpec.scala:41)
[info]   at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
[info]   at org.scalatest.Transformer.org$scalatest$Transformer$$$anonfun$1(Transformer.scala:22)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:22)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:20)
[info]   at org.scalatest.WordSpecLike$$anon$1.apply(WordSpecLike.scala:953)
[info]   at org.scalatest.Suite.withFixture(Suite.scala:1122)
[info]   at org.scalatest.WordSpec.withFixture(WordSpec.scala:1881)
[info]   at org.scalatest.WordSpecLike.invokeWithFixture$1(WordSpecLike.scala:951)
[info]   at org.scalatest.WordSpecLike.org$scalatest$WordSpecLike$$$anonfun$5(WordSpecLike.scala:962)
[info]   at org.scalatest.SuperEngine.runTestImpl(Engine.scala:306)
[info]   at org.scalatest.WordSpecLike.runTest(WordSpecLike.scala:962)
[info]   at org.scalatest.WordSpec.runTest(WordSpec.scala:1881)
[info]   at org.scalatest.WordSpecLike.org$scalatest$WordSpecLike$$$anonfun$6(WordSpecLike.scala:1021)
[info]   at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$$anonfun$8(Engine.scala:413)
[info]   at scala.collection.immutable.List.foreach(List.scala:376)
[info]   at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)
[info]   at org.scalatest.SuperEngine.runTestsInBranch(Engine.scala:390)
[info]   at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$$anonfun$8(Engine.scala:427)
[info]   at scala.collection.immutable.List.foreach(List.scala:376)
[info]   at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)
[info]   at org.scalatest.SuperEngine.runTestsInBranch(Engine.scala:396)
[info]   at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:483)
[info]   at org.scalatest.WordSpecLike.runTests(WordSpecLike.scala:1021)
[info]   at org.scalatest.WordSpec.runTests(WordSpec.scala:1881)
[info]   at org.scalatest.Suite.run(Suite.scala:1424)
[info]   at org.scalatest.WordSpec.org$scalatest$WordSpecLike$$super$run(WordSpec.scala:1881)
[info]   at org.scalatest.WordSpecLike.org$scalatest$WordSpecLike$$$anonfun$7(WordSpecLike.scala:1067)
[info]   at org.scalatest.SuperEngine.runImpl(Engine.scala:545)
[info]   at org.scalatest.WordSpecLike.run(WordSpecLike.scala:1067)
[info]   at org.scalatest.WordSpec.run(WordSpec.scala:1881)
[info]   at org.scalatest.tools.Framework.org$scalatest$tools$Framework$$runSuite(Framework.scala:357)
[info]   at org.scalatest.tools.Framework$ScalaTestTask.execute(Framework.scala:529)
[info]   at sbt.TestRunner.runTest$1(TestFramework.scala:76)
[info]   at sbt.TestRunner.run(TestFramework.scala:85)
[info]   at sbt.TestFramework$$anon$2$$anonfun$$init$$1$$anonfun$apply$8.apply(TestFramework.scala:202)
[info]   at sbt.TestFramework$$anon$2$$anonfun$$init$$1$$anonfun$apply$8.apply(TestFramework.scala:202)
[info]   at sbt.TestFramework$.sbt$TestFramework$$withContextLoader(TestFramework.scala:185)
[info]   at sbt.TestFramework$$anon$2$$anonfun$$init$$1.apply(TestFramework.scala:202)
[info]   at sbt.TestFramework$$anon$2$$anonfun$$init$$1.apply(TestFramework.scala:202)
[info]   at sbt.TestFunction.apply(TestFramework.scala:207)
[info]   at sbt.Tests$$anonfun$9.apply(Tests.scala:216)
[info]   at sbt.Tests$$anonfun$9.apply(Tests.scala:216)
[info]   at sbt.std.Transform$$anon$3$$anonfun$apply$2.apply(System.scala:44)
[info]   at sbt.std.Transform$$anon$3$$anonfun$apply$2.apply(System.scala:44)
[info]   at sbt.std.Transform$$anon$4.work(System.scala:63)
[info]   at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:226)
[info]   at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:226)
[info]   at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
[info]   at sbt.Execute.work(Execute.scala:235)
[info]   at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:226)
[info]   at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:226)
[info]   at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
[info]   at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
[info]   at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[info]   at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[info]   at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[info]   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
[info]   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
[info]   at java.lang.Thread.run(Thread.java:745)
[info] - should handle outers with constructors
[info] - should Handle functions in traits
[info] - should Handle captured vals
```

I'll see if I can manage to get a clearer stacktrace in order to figure out what's failing inside of the shaded ASM.

While Spark doesn't depend on Kryo's ClosureCleaner, it's likely that Spark's own ClosureCleaner (which is where Kryo's originated) has similar problems that will need to be addressed, so I'd like to try to take the chance to fix those problems here, too.

**Comment by JoshRosen** [comment]
Aha, I spot the problem: I think we're going to have to port that code to ASM 5. I did a similar port in Spark at https://github.com/apache/spark/pull/9512, so I'll re-apply those changes here.

**Comment by johnynek** [comment]
Thanks for pushing on this, @JoshRosen

**Comment by magro** [comment]
We just released kryo 4.0.0 btw.

**Comment by johnynek** [comment]
@magro sadly, the transitive dependencies are such that kryo is very hard to upgrade (chill <- storm <- summingbird) and (chill <- scalding <- summingbird) means that coordinating with apache storm releases, which is a bit of a pain.

**Comment by magro** [comment]
@johnynek I understand. Do you see chill as the (intended) standard lib for adding scala support to kryo? I usually recommend it like this, but perhaps the focus is more on adding support for Hadoop, storm, scalding etc? I'm wondering if it's worth to have a dedicated project for adding kryo scala support that can keep up with the latest kryo release without having to worry about compatibility with the mentioned projects. WDYT? (we could also discuss this in the kryo Gitter channel)

**Comment by johnynek** [comment]
well, maybe we should just start having branches. I don't want to force the world to be blocked on storm's choices. They should shade or isolate the classpaths.

We can upgrade to `4.0.0`. We'll just publish older versions, I suppose.

**Comment by magro** [comment]
Ok, sounds good. 

I'm currently wondering if kryo should provide serializers for Scala, e.g. published as kryo-scala artifact. Would you be interested in this, do you think it would be used by chill?

**Comment by johnynek** [comment]
Yes, in general I think it would be great to move more of chill directly to
kryo.

**Comment by magro** [comment]
Ok. Do you have more things in mind than just scala serializers?

**Comment by johnynek** [comment]
most of this is super useful:

https://github.com/twitter/chill/tree/develop/chill-java/src/main/java/com/twitter/chill

mostly because in many places where you want kryo, you also want to send how to instantiate the kryo instances, and perhaps share instances between threads.

**Comment by magro** [comment]
Ah ok. There is a related PR in kryo: https://github.com/EsotericSoftware/kryo/pull/339
I'd have to take a closer look at chill's solution to see how it compares. At least it's a production proven solution... 

Re sharing instances: kryo now also provides a pool - https://github.com/EsotericSoftware/kryo/blob/master/src/com/esotericsoftware/kryo/pool/KryoPoolQueueImpl.java

**Comment by magro** [comment]
For the kryo-scala artifact I just submitted https://github.com/EsotericSoftware/kryo/issues/452.

**Comment by johnynek** [comment]
cool. sharing kryo-scala would be nice.

**Comment by sritchie** [comment]
Closed, since this is handled by #266 and #271 . Thanks!
